estimagic/logging/create_database.py

```diff
@@ -0,0 +1,238 @@
+        accessed via ``metadata.bind``.
+
+    """
+    if isinstance(path, str):
+        path = Path(path)
+
+    if path.exists() and replace:
+        path.unlink()
+
+    engine = create_engine(f"sqlite:///{path}")
+    _make_engine_thread_safe(engine)
+    metadata = MetaData()
+    metadata.bind = engine
+    metadata.reflect()
+    return metadata
+
+
+def _make_engine_thread_safe(engine):
+    """Make the engine even more thread safe than by default.
+
+    The code is taken from the documentation: https://tinyurl.com/u9xea5z
+
+    The main purpose is to emit the begin statement of any connection
+    as late as possible in order to keep the time in which the database
+    is locked as short as possible.
+
+    """
+
+    @event.listens_for(engine, "connect")
+    def do_connect(dbapi_connection, connection_record):
+        # disable pysqlite's emitting of the BEGIN statement entirely.
+        # also stops it from emitting COMMIT before absolutely necessary.
+        dbapi_connection.isolation_level = None
+
+    @event.listens_for(engine, "begin")
+    def do_begin(conn):
+        # emit our own BEGIN
+        conn.execute("BEGIN DEFERRED")
+
+
+def prepare_database(
+    path,
+    params,
+    db_options=None,
+    optimization_status="scheduled",
+    gradient_status=0,
+    replace=False,
+):
+    """Return database metadata object with all relevant tables.
+
+    This should always be used to create entirely new databases or to create the
+    tables needed during optimization in an existing database.
+
+    The resulting database has the following tables:
+    - params_history: the complete history of parameters from the optimization. The
+      first column is "iteration", the remaining columns are parameter names taken
+      from params["name"].
+    - gradient_history: the complete history of gradient evaluations from the
+      optimization. Same columns as params_history.
+    - criterion_history: the complete history of criterion values from the optimization.
+      The first column is "iteration", the second column is "value".
+    - time_stamps: timestamps from the end of each criterion evaluation. Same columss as
+      criterion_history.
+    - convergence_history: the complete history of convergence criteria from the
+      optimization. The first column is iteration, the other columns are "ftol", "gtol"
+      and "xtol".
+    - start_params: copy of user provided params.
+    - optimization_status: table with one row and one column called "value" which takes
+      the values "scheduled", "running", "success" or "failure". The initial value is
+      "scheduled".
+    - gradient_status: table with one row and one column called "value" which takes
+      can be any float between 0 and 1 and indicates the progress of the gradient
+      calculation. Initialized to 0.
+    - db_options: table with one row and one column called "value". It contains
```

Comment:
If this can still be changed, I would go for the more verbose `dashboard_options` everywhere. To me, `db` is short for database; that had always confused me when using estimagic and it only becomes worse now that there actually is a database backend in it...

Reply:
Yes, I realized the same thing when I worked on this PR. But I would do it in a separate PR because we it has to be done in many places! I created #98 to discuss solutions.